lib/diff.ml

```diff
@@ -194,13 +188,31 @@ let rec type_expr ~typing_env ?(ref_params = []) ?(cur_params = []) reference
       | Stddiff.Same _ -> Stddiff.Same reference
       | Changed change -> Changed (Arrow change))
   | Tconstr (ref_path, ref_args, _), Tconstr (cur_path, cur_args, _) -> (
-      let constr =
-        constr ~typing_env ~ref_params ~cur_params
-          ~reference:(ref_path, ref_args) ~current:(cur_path, cur_args)
+      let expanded_ref =
+        Typing_env.fully_expand_type_expr ~typing_env ~type_expr:reference
+          ~path:ref_path ~args:ref_args
+      in
+      let expanded_cur =
+        Typing_env.fully_expand_type_expr ~typing_env ~type_expr:current
+          ~path:cur_path ~args:cur_args
       in
-      match constr with
-      | Stddiff.Same _ -> Stddiff.Same current
-      | Changed change -> Changed (Constr change))
+      match (Types.get_desc expanded_ref, Types.get_desc expanded_cur) with
+      | Tconstr (ref_path, ref_args, _), Tconstr (cur_path, cur_args, _) -> (
+          let constr =
+            constr ~typing_env ~ref_params ~cur_params
+              ~reference:(ref_path, ref_args) ~current:(cur_path, cur_args)
+          in
+          match constr with
+          | Stddiff.Same _ -> Stddiff.Same current
+          | Changed change -> Changed change)
+      | _, _ -> (
+          let diff =
+            type_expr ~typing_env ~ref_params ~cur_params expanded_ref
+              expanded_cur
+          in
+          match diff with
+          | Same _ -> Same current
+          | Changed change -> Changed change))
```

Comment:
Let's move this into an `expand_and_diff_tconstr` to keep the type_expr function a bit more readable and reasonably sized.